src/server/server.ts

```diff
@@ -242,6 +287,183 @@ export class CTProtoServer<AuthRequestPayload, AuthData, ApiRequest extends NewM
+       */
+      if (chunkNumber === 0) {
+        this.uploadingFiles.push( { id: fileId,
+          uploadedChunks: [],
+          file: fileData,
+          chunks: payload.chunks,
+          payload: payload.payload,
+          type: payload.type,
+        } );
+      } else {
+        this.uploadingFiles.push( { id: fileId,
+          uploadedChunks: [],
+          file: fileData,
+        } );
+      }
+
+      file = this.uploadingFiles.find((req) => req.id === fileId);
+    } else {
+      /**
+       * Clear timeout, if chunk comes
+       */
+      if (file.uploadingWaitingTimeoutId) {
+        clearTimeout(file.uploadingWaitingTimeoutId);
+      }
+
+      /**
+       * In case if the first chunk comes not first, push more file data to uploading file object
+       */
+      if (chunkNumber === 0) {
+        file.chunks = payload.chunks;
+        file.type = payload.type;
+        file.payload = payload.payload;
+      }
+
+      /**
+       * Update file data
+       */
+      this.updateFileData(file, chunkOffset, fileChunk);
+    }
+
+    if ( !file ) {
+      return;
+    }
+
+    file.uploadedChunks[chunkNumber] = true;
+
+    if (this.isFileFullyUploaded(file)) {
+      /**
+       * Make file request object
+       */
+      const parsedFile = {
+        type: file.type,
+        payload: file.payload,
+        file: file.file,
+      } as ApiUploadRequest;
+
+      /**
+       * Make response for fully uploaded file
+       */
+      const response = await this.options.onUploadMessage(parsedFile);
+
+      client.respond(fileId, response);
+    } else {
+      /**
+       * Make response for incoming chunk
+       */
+      const response = {
+        chunkNumber: chunkNumber,
+        type: file.type,
+        fileId: file.id,
+      };
+
+      client.respond(payload.id, response);
+    }
+
+    /**
+     * Set timeout to remove file, in case of no chunks incoming
+     */
+    file.uploadingWaitingTimeoutId = setTimeout( () => {
+      if (file) {
+        this.uploadingFiles.splice(this.uploadingFiles.indexOf(file), 1);
+      }
+    }, this.chunkWaitingTimeout );
+  }
+
+  /**
+   * Update file data, insert incoming chunk
+   *
+   * @param file - uploading file
+   * @param chunkSlice - chunk offset
+   * @param fileChunk - data to insert
+   */
+  private updateFileData(file: UploadingFile, chunkSlice: number, fileChunk: Buffer): void {
+    let fileData;
+
+    /**
+     * Push file data
+     */
+    if (file.file.length < chunkSlice + fileChunk.length) {
+      fileData = Buffer.alloc(chunkSlice + fileChunk.length);
+      file.file.copy(fileData);
+      fileChunk.copy(fileData, chunkSlice);
+    } else {
+      fileData = file.file;
+      fileChunk.copy(fileData, chunkSlice);
+    }
+
+    file.file = fileData;
+  }
+
+  /**
+   * Check is file fully uploaded
+   *
+   * @param file - uploading file
+   */
+  private isFileFullyUploaded(file: UploadingFile): boolean {
```

Comment:
could we move all the methods working with file uploading to the separate class? to not to make the Server class too big